### PR TITLE
Merge mixed mesh work into scorec

### PR DESCRIPTION
**Pull request by joshia5**

### CODE REVIEW

**Comment by joshia5** [comment]
following tests failed :
run_unit_io (Failed) -- segmentation fault
run_mpi_unit_io (Failed)-- segmentation fault
warp_test_parallel (Failed) -- max diff at vertex 1008, comp 0, values 6.327799822371828e+01 vs 6.327799822371829e+01
vertex tag "metric" values are different
max diff at edge 6073, comp 0, values 1.124971094950606e+00 vs 1.124971094950607e+00
edge tag "length" values are different
tet tag "quality" values are different
max diff at region 3483, comp 0, values 7.559526299369257e-01 vs 7.559526299369259e-01
This run, stored at "gold_warp_bad.osh",
does not match the gold at "gold_warp.osh"
<end of output>
Test time = 37.16 sec

**Comment by joshia5** [comment]
error in osh and/or vtu/vtk IO APIs

**Comment by cwsmith** [comment]
Do all tests pass in `master` prior to the merge?

**Comment by joshia5** [comment]
The IO tests pass but warp_test_parallel fails when building omegaH from master : 

test took 35.3119 seconds
max diff at vertex 1008, comp 0, values 6.327799822371828e+01 vs 6.327799822371829e+01
vertex tag "metric" values are different
max diff at edge 6073, comp 0, values 1.124971094950606e+00 vs 1.124971094950607e+00
edge tag "length" values are different
max diff at region 3483, comp 0, values 7.559526299369257e-01 vs 7.559526299369259e-01
tet tag "quality" values are different
This run, stored at "gold_warp_bad.osh",
does not match the gold at "gold_warp.osh"
<end of output>
Test time =  37.08 sec
Test Failed.

**Comment by joshia5** [comment]
On building using flag -DCMAKE_CUDA_FLAGS="-arch=sm_75" flag instead of -DOmega_h_ARCH=sm_75, all tests pass except mixed_test which fails with error :
Omega_h_array.hpp:51: int Omega_h::Write<T>::size     () const [with T = signed char]: block: [0,0,0], thread: [0,0,0] Assertion `exists()` failed.
 162 terminate called after throwing an instance of 'std::runtime_error'
 163   what():  cudaDeviceSynchronize() error( cudaErrorAssert): device-side assert triggered /lore/joshia5/     develop/merge_testing/mixed/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:120
 164 Traceback functionality not available

This indicates an error with build process

**Comment by cwsmith** [comment]
@joshia5 Is this with Simmetrix support enabled?

**Comment by joshia5** [comment]
yes

**Comment by cwsmith** [comment]
@joshia5 Sounds good.  Thanks for the update.  We should also test with Simmetrix disabled before merging.

File: src/Omega_h_array_ops.cpp
Comment (cwsmith) [inline src/Omega_h_array_ops.cpp:141]: I think this may run faster as two kernels without the conditional.

File: src/Omega_h_vtk.cpp
Comment (cwsmith) [inline src/Omega_h_vtk.cpp:63]: please remove commented code

**Review by cwsmith — APPROVED** [review]
The code looks clean.  I think the `concat` function could be split into two kernels for slightly (?) better performance... It is your call if you want to make the change.

**Comment by cwsmith** [comment]
| all tests pass apart from the warp_test (same behavior as scorec master) 

@joshia5 Thank you for testing.

File: src/Omega_h_vtk.cpp
Comment (joshia5) [reply src/Omega_h_vtk.cpp]: resolved here https://github.com/SCOREC/omega_h/pull/5/commits/bee665e7840480f114ea95b6ca521591bfc4fbb5

File: src/Omega_h_array_ops.cpp
Comment (joshia5) [reply src/Omega_h_array_ops.cpp:141]: resolved here https://github.com/SCOREC/omega_h/pull/5/commits/bee665e7840480f114ea95b6ca521591bfc4fbb5

**Comment by joshia5** [comment]
@cwsmith  a thing to note here is that the scorec/master branch cannot be used as a converter for simmetrix to omega_h mesh conversion. The converter (available only for simplex meshes) should be called from this branch https://github.com/SCOREC/omega_h/tree/cws/meshSimReader

**Pull request merged**
